m3/src/reporter_impl.cc

```diff
@@ -0,0 +1,438 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#include "m3/src/reporter_impl.h"
+
+#include <algorithm>
+#include <cmath>
+#include <iostream>
+#include <limits>
+#include <sstream>
+
+#include "boost/format.hpp"
+#include "thrift/protocol/TCompactProtocol.h"
+
+#include "m3/reporter.h"
+#include "m3/src/calc_transport.h"
+#include "m3/thrift/M3.h"
+#include "tally/src/capable_of.h"
+
+using apache::thrift::protocol::TCompactProtocol;
+using apache::thrift::transport::TTransportException;
+
+namespace tally {
+
+namespace m3 {
+
+const char Reporter::Impl::HISTOGRAM_BUCKET_NAME[] = "bucket";
+const char Reporter::Impl::HISTOGRAM_BUCKET_ID_NAME[] = "bucketid";
+
+Reporter::Impl::Impl(const std::string &host, uint16_t port,
+                     const std::map<std::string, std::string> &common_tags,
+                     uint32_t max_queue_size, uint16_t max_packet_size)
+    : common_tags_(ConvertTags(common_tags)),
+      max_queue_size_(max_queue_size),
+      // Reserve 20% of the packet size for encoding overhead.
```

Comment:
You shouldn't need to do this if we're using the Calculate transport?  We know exactly when we overflow a packet and don't need to any soft maths.

Reply:
I'm using the Calculate Transport I'm just not using in a strict matter. I add each metric to the running total but I don't calculate the size of the batch each time so this was just to give us some slack space. If I calculate the encoding size of the batch each time I could calculate the size directly but it seems like a waste to keep re-encoding the same metrics until we hit the limit.

Reply:
What we do in the Jaeger reporter is just reserve 65000 bytes which is max packet for UDP on most platforms anyway.